crates/volta-core/src/tool/yarn/resolve.rs

```diff
@@ -8,56 +12,90 @@ use crate::session::Session;
 use crate::style::progress_spinner;
 use crate::tool::Yarn;
 use crate::version::{parse_version, VersionSpec, VersionTag};
+use attohttpc::header::ACCEPT;
 use attohttpc::Response;
-use cfg_if::cfg_if;
 use log::debug;
 use semver::{Version, VersionReq};
 
-// ISSUE (#86): Move public repository URLs to config file
-cfg_if! {
-    if #[cfg(feature = "mock-network")] {
-        fn public_yarn_version_index() -> String {
-            format!("{}/yarn-releases/index.json", mockito::SERVER_URL)
-        }
-        fn public_yarn_latest_version() -> String {
-            format!("{}/yarn-latest", mockito::SERVER_URL)
-        }
-    } else {
-        /// Return the URL of the index of available Yarn versions on the public git repository.
-        fn public_yarn_version_index() -> String {
-            "https://api.github.com/repos/yarnpkg/yarn/releases".to_string()
-        }
-        /// URL of the latest Yarn version on the public yarnpkg.com
-        fn public_yarn_latest_version() -> String {
-            "https://yarnpkg.com/latest-version".to_string()
-        }
-    }
-}
-
 pub fn resolve(matching: VersionSpec, session: &mut Session) -> Fallible<Version> {
     let hooks = session.hooks()?.yarn();
     match matching {
         VersionSpec::Semver(requirement) => resolve_semver(requirement, hooks),
         VersionSpec::Exact(version) => Ok(version),
-        VersionSpec::None | VersionSpec::Tag(VersionTag::Latest) => resolve_latest(hooks),
-        VersionSpec::Tag(tag) => Err(ErrorKind::YarnVersionNotFound {
+        VersionSpec::None => resolve_tag(VersionTag::Latest, hooks),
+        VersionSpec::Tag(tag) => resolve_tag(tag, hooks),
+    }
+}
+
+fn resolve_tag(tag: VersionTag, hooks: Option<&ToolHooks<Yarn>>) -> Fallible<Version> {
+    // This triage is complicated because we need to maintain the legacy behavior of hooks
+    // First, if the tag is 'latest' and we have a 'latest' hook, we use the old behavior
+    // Next, if the tag is 'latest' and we _do not_ have a 'latest' hook, we use the new behavior
+    // Next, if the tag is _not_ 'latest' and we have an 'index' hook, we show an error since
+    //     the previous behavior did not support generic tags
+    // Finally, we don't have any relevant hooks, so we can use the new behavior
```

Comment:
I totally get wanting to maintain legacy behavior and avoid introducing surprises. However, it seems inconsistent to add support for the new behavior of `'latest'` but *not* other tags—what's the failure mode there? (I assume there is one that just isn't obvious to me! 😅)

Reply:
It may be that the comment isn't clear, the goal is that if you have hooks set, we use the old behavior (which _only_ supported the `latest` tag and no others), while if you don't have hooks, then we use the new logic. The cases are:

1. We have a `latest` hook and are looking up the `latest` tag: This was supported under the old system, so we can use the legacy logic for resolving latest.
2. We are looking up the `latest` tag but _don't_ have the `latest` hook: Previously, that would use the default, public registry, so we can port that over to use the new latest logic.
3. We are looking up a custom tag and we have an `index` hook: This was _not_ supported previously. Since we'd be using the `index` hook to look up data in the GitHub API format, there would be no way to resolve a tag anyway, so we can fail fast.
4. Last if we don't have an `index` hook either, then we are using the new default behavior from `npm` (the most common case).

We could potentially change it to have a slightly clearer break:

If you have any hooks defined (i.e. if the `hooks` parameter is `Some(_)`), then we always use the old logic, regardless of whether we actually have a relevant hook or not. If you have no hooks then we use the new logic.

That would simplify the split here, but would also mean that adding an `index` hook would change the behavior of `volta install yarn@latest`, which seems like it would introduce more confusion.

Reply:
This makes sense enough to me to proceed. My *only* remaining question is: why not just use the *new* behavior for everything (which I think would simplify things) and mark it as a breaking change—perhaps with a feature flag approach, doing a release which only supports the old way but deprecates it, then a bit later enabling the feature flag and removing the old way? (Not a blocker by any stretch.)